### PR TITLE
READY : tail

### PR DESCRIPTION
**Pull request by SupperZum**

### CODE REVIEW

File: text/src/tail.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    let start = if n < 0 {
+        (lines.len() as isize + n).max(0) as usize
+    } else {
+        (n - 1).max(0) as usize
+    };
+
+    if start > lines.len() {
+        print!("");
+        return Ok(());
+    }
```
Comment (0x07C0) [inline text/src/tail.rs:67]: This is invalid. If the file has only `x` lines and we want to print `y` we should print at least those `x` lines

File: text/src/tail.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        .expect("Failed to read file");
+    let start = if n < 0 {
+        (buffer.len() as isize + n).max(0) as usize
+    } else {
+        (n - 1).max(0) as usize
+    };
+    if start > buffer.len() {
+        print!("");
+        return Ok(());
+    }
```
Comment (0x07C0) [inline text/src/tail.rs:95]: Same as with lines. Don't just print nothing if the file is smaller then we need to print.

File: text/src/tail.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        } else {
+            Box::new(fs::File::open(args.file.as_ref().unwrap())?)
+        }
+    };
+    let mut reader = io::BufReader::new(file);
+
+    match args.bytes {
+        Some(bytes) => print_last_n_bytes(&mut reader, bytes)?,
+        None => print_last_n_lines(reader, args.lines.unwrap())?,
+    }
```
Comment (0x07C0) [inline text/src/tail.rs:123]: Weird use of match. `if let ... else if let` could look better here to get rid of the unwrap

File: text/src/tail.rs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+        loop {
+            let mut buffer = String::new();
+            let bytes_read = reader.read_line(&mut buffer)?;
+            if bytes_read > 0 {
+                print!("{}", buffer);
+                io::stdout().flush()?;
+            }
+            thread::sleep(Duration::from_millis(60));
+        }
```
Comment (0x07C0) [inline text/src/tail.rs:142]: use inotify. sleeping is very sub-optimal.

File: text/src/tail.rs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    match args.bytes {
+        Some(bytes) => print_last_n_bytes(&mut reader, bytes)?,
+        None => print_last_n_lines(reader, args.lines.unwrap())?,
+    }
+
+    // If follow option is specified, continue monitoring the file
+    if args.follow {
+        let file_path = args.file.as_ref().unwrap();
+        let mut file = fs::File::open(file_path)?;
```
Comment (0x07C0) [inline text/src/tail.rs:128]: We should also be able to follow stdin

File: text/src/tail.rs (hunk 1 above)
Comment (SupperZum) [reply text/src/tail.rs:67]: I focused on the test cases in the original repository. 
Example: https://github.com/coreutils/coreutils/blob/master/tests/tail/tail.pl#L31

File: text/src/tail.rs (hunk 5 above)
Comment (SupperZum) [reply text/src/tail.rs:128]: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/tail.html
"If no file operand is specified and standard input is a pipe or FIFO, the -f option shall be ignored."
It seems to me that this option is not considered according to the description.

**Pull request merged**